plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -68,46 +105,108 @@ class bp_connection_manager {
+            EOS_ASSERT(false, chain::plugin_config_exception,
+                       "the account ${a} supplied by --p2p-auto-bp-peer option is invalid", ("a", entry));
          }
-      } catch (eosio::chain::name_type_exception&) {
-         EOS_ASSERT(false, chain::plugin_config_exception, "the account supplied by --auto-bp-peer option is invalid");
       }
    }
 
    // Only called at plugin startup
    template <typename T>
    void for_each_bp_peer_address(T&& fun) const {
-      for (auto& [_, addr] : config.bp_peer_addresses) { fun(addr); }
+      fc::lock_guard g(gossip_bps.mtx);
+      for (const auto& bp_peer : gossip_bps.index) {
+         fun(bp_peer.server_address);
+      }
+   }
+
+   // Only called at plugin startup
+   void set_bp_producer_peers(const std::vector<std::string>& peers) {
+      for (const auto& entry : peers) {
+         try {
+            config.my_bp_peer_accounts.emplace(chain::name(entry));
+         } catch (chain::name_type_exception&) {
+            EOS_ASSERT(false, chain::plugin_config_exception,
+                       "the producer ${p} supplied by --p2p-producer-peer option is invalid", ("p", entry));
+         }
+      }
+   }
+
+   // Called at startup and when peer key changes
+   // empty modified_keys means to update all
+   void update_bp_producer_peers(const chain::controller& cc, const flat_set<account_name>& modified_keys, const std::string& server_address)
+   {
+      if (config.my_bp_peer_accounts.empty())
+         return;
+      fc::lock_guard gm(mtx);
+      fc::lock_guard g(gossip_bps.mtx);
+      // normally only one bp peer account except in testing scenarios or test chains
+      for (const auto& e : config.my_bp_peer_accounts) { // my_bp_peer_accounts not modified after plugin startup
+         if (modified_keys.empty() || modified_keys.contains(e)) {
+            std::optional<public_key_type> pk = cc.get_peer_key(e);
+            // EOS_ASSERT can only be hit on plugin startup, otherwise this method called with modified_keys that are in cc.get_peer_key()
+            EOS_ASSERT(pk, chain::plugin_config_exception, "No on-chain peer key found for ${n}", ("n", e));
+            fc_dlog(self()->get_logger(), "Signing with producer_name ${p} key ${k}", ("p", e)("k", *pk));
+            if (e == *config.my_bp_peer_accounts.begin()) { // use the first one of the set, doesn't matter which one is used, just need one
+               gossip_bp_peers_message::bp_peer signed_empty{.producer_name = e}; // .server_address not set for initial message
+               signed_empty.sig = self()->sign_compact(*pk, signed_empty.digest());
+               initial_gossip_msg_factory.set_initial_send_buffer(signed_empty);
+            }
+            auto& prod_idx = gossip_bps.index.get<by_producer>();
+            gossip_bp_peers_message::bp_peer peer{.producer_name = e, .server_address = server_address};
+            peer.sig = self()->sign_compact(*pk, peer.digest());
```

Comment:
Are we not signing with our own producer key here? If so why do we need to check `modified_keys.contains(e)`? Wouldn't we have restarted our node if we changed our registered key on-chain?

Reply:
Should we check that `sign_compact` does find the private key associated to `*pk`, and that the signature stored in `peer.sig` is not empty?